internal/action/actions.go

```diff
@@ -160,6 +160,50 @@ func (h *BufPane) Center() bool {
 	return true
 }
 
+// CursorToViewTop moves the cursor to the top of the view,
+// offset by scrollmargin unless at the beginning or end of the file
+func (h *BufPane) CursorToViewTop() bool {
+	v := h.GetView()
+	h.Buf.ClearCursors()
+	scrollmargin := int(h.Buf.Settings["scrollmargin"].(float64))
+	if v.StartLine.Line < scrollmargin {
```

Comment:
Again, this doesn't take softwrap into account. A line may be so long that it doesn't fit in the entire view, so the cursor may be in the very first line of the file, yet still not at the top of the file.

See `Relocate()` in `internal/display/bufwindow.go` for an example of how to do such checks correctly.

Reply:
I'm not sure how the current cursor location is relevant here, since we're moving the cursor regardless of its current location. In the case of `CursorToViewTop` we want to place the cursor at the very top of the current view, regardless of whether that top is in the middle of a softwrapped line or not. And with `CursorToViewBottom` we want to place the cursor at the beginning of the last line in the view, regardless of whether that line is the very end of the file, or in the middle of a softwrapped line.

Am I missing something?

Reply:
I'm not talking about the current cursor location, I'm talking about the location where we are going to move the cursor to. I'm talking about the current view location and your `v.StartLine.Line < scrollmargin` check.

Once again, using `StartLine.Line` is incorrect, since `StartLine.Line` is the number of the logical line, while we should be dealing with visual lines here. That's what `SLoc` is for.

Try turning `softwrap` on, adding a very long as the first line in the buffer, slightly scrolling the view down so that the beginning of this first line is outside the view, then triggering `CursorToViewTop`, and then moving the cursor with an arrow key. The view will jump a few lines down due to scrollmargin, i.e. the very problem you've been fighting against is back.

Correct way is for example: `if v.StartLine.LessThan(h.Scroll(display.SLoc{0, 0}, scrollmargin))`.

Reply:
OK, that makes sense. Fixed (I think) and pushed.